src/Microsoft.Azure.NotificationHubs.DotNetCore.Tests/NotificationHubClientTest.cs

```diff
@@ -908,6 +908,76 @@ private async Task ScheduleNotificationAsync_SendAppleNativeNotification_GetSucc
+
+            Assert.IsType<FcmRegistrationDescription>(receivedRegistration);
+            RecordTestResults();
+        }
+
+        [Fact]
+        private async Task GetRegistrationAsync_CreateFcmNativeRegistrationThenGetFcmRegistrationDescription_GetFcmRegistrationType()
+        {
+            LoadMockData();
+            await DeleteAllRegistrationsAndInstallations();
+
+            var registration = new FcmRegistrationDescription(_configuration["GcmDeviceToken"]);
+            var createdRegistration = await _hubClient.CreateRegistrationAsync(registration);
+            var receivedRegistration = await _hubClient.GetRegistrationAsync<FcmRegistrationDescription>(createdRegistration.RegistrationId);
+
+            Assert.IsType<FcmRegistrationDescription>(receivedRegistration);
+            RecordTestResults();
+        }
+
+        [Fact]
+        private async Task GetRegistrationAsync_CreateFcmTemplateRegistrationThenGetRegistrationDescription_GetFcmTemplateRegistrationType()
+        {
+            LoadMockData();
+            await DeleteAllRegistrationsAndInstallations();
+
+            var registration = new FcmTemplateRegistrationDescription(_configuration["GcmDeviceToken"], "{\"data\":{\"message\":\"Message\"}}");
+            registration.PushVariables = new Dictionary<string, string>()
+            {
+                {"var1", "value1"}
+            };
+            registration.Tags = new HashSet<string>() { "tag1" };
+            registration.TemplateName = "Template Name";
+
+            var createdRegistration = await _hubClient.CreateRegistrationAsync(registration);
+            var receivedRegistration = await _hubClient.GetRegistrationAsync<RegistrationDescription>(createdRegistration.RegistrationId);
+            
+            Assert.IsType<FcmTemplateRegistrationDescription>(receivedRegistration);
+            RecordTestResults();
+        }
+
+        [Fact]
+        private async Task GetRegistrationAsync_CreateFcmTemplateRegistrationThenGetFcmTemplateRegistrationDescription_GetFcmTemplateRegistrationType()
+        {
+            LoadMockData();
+            await DeleteAllRegistrationsAndInstallations();
+
+            var registration = new FcmTemplateRegistrationDescription(_configuration["GcmDeviceToken"], "{\"data\":{\"message\":\"Message\"}}");
+            registration.PushVariables = new Dictionary<string, string>()
+            {
+                {"var1", "value1"}
+            };
+            registration.Tags = new HashSet<string>() { "tag1" };
+            registration.TemplateName = "Template Name";
+
+            var createdRegistration = await _hubClient.CreateRegistrationAsync(registration);
+            var receivedRegistration = await _hubClient.GetRegistrationAsync<FcmTemplateRegistrationDescription>(createdRegistration.RegistrationId);
+
+            Assert.IsType<FcmTemplateRegistrationDescription>(receivedRegistration);
+            RecordTestResults();
+        }
```

Comment:
I would like to add 2 more test scenarios for `GetRegistrationAsync`:

- `GetRegistrationAsync<GcmRegistrationDescription>()` returns `GcmRegistrationDescription`.
- `GetRegistrationAsync<GcmTemplateRegistrationDescription>()` returns `GcmTemplateRegistrationDescription`.

We need to make sure all of the Registration APIs behave in the same manner.

- When creating or updating a registration, the return type must be the same as the type passed in.
  - `CreateRegistrationAsync(new FcmRegistrationDescription())` returns an instance of `FcmRegistrationDescription`.
  - `CreateRegistrationAsync(new GcmRegistrationDescription())` returns an instance of `GcmRegistrationDescription`.
  - same applies for the `Template` types.
  - same applies for the `Update` methods.
- When listing registrations (Get All, Get by Channel, Get by Tag) we should continue to map `Gcm` to `Fcm` registrations. This is the current behavior, but our test coverage is weak. Let's add more tests to ensure we are doing the right thing.
- When getting a single registration, the return type should match the generic type parameter.
  - `GetRegistrationAsync<FcmRegistrationDescription>()` returns an instance of `FcmRegistrationDescription`.
  - `GetRegistrationAsync<GcmRegistrationDescription>()` returns an instance of `GcmRegistrationDescription`.
  - `GetRegistrationAsync<RegistrationDescription>()` returns an instance of `FcmRegistrationDescription` (since the `Fcm` type is the "default").
  - same applies to the `Template` types.

Reply:
@brannon have added tests for check that:

GetRegistrationAsync:
- `GetRegistrationAsync<GcmRegistrationDescription>()` returns `GcmRegistrationDescription`
- `GetRegistrationAsync<GcmTemplateRegistrationDescription>()` returns `GcmTemplateRegistrationDescription`
- `GetRegistrationAsync<FcmRegistrationDescription>()` returns `FcmRegistrationDescription`
- `GetRegistrationAsync<FcmTemplateRegistrationDescription>()` returns `FcmTemplateRegistrationDescription`
- `GetRegistrationAsync<RegistrationDescription>()` returns `FcmRegistrationDescription`
- `GetRegistrationAsync<RegistrationDescription>()` returns `FcmTemplateRegistrationDescription`

CreateRegistrationAsync:

- `CreateRegistrationAsync(new FcmRegistrationDescription())` returns an instance of `FcmRegistrationDescription`
- `CreateRegistrationAsync(new FcmTemplateRegistrationDescription())` returns an instance of `FcmTemplateRegistrationDescription`
- `CreateRegistrationAsync(new GcmRegistrationDescription())` returns an instance of `GcmRegistrationDescription`
- `CreateRegistrationAsync(new GcmTemplateRegistrationDescription())` returns an instance of `GcmTemplateRegistrationDescription`

CreateOrUpdateRegistrationAsync:
- `CreateOrUpdateRegistrationAsync(new FcmRegistrationDescription())` returns an instance of `FcmRegistrationDescription`
- `CreateOrUpdateRegistrationAsync(new FcmTemplateRegistrationDescription())` returns an instance of `FcmTemplateRegistrationDescription`
- `CreateOrUpdateRegistrationAsync(new GcmRegistrationDescription())` returns an instance of `GcmRegistrationDescription`
- `CreateOrUpdateRegistrationAsync(new GcmTemplateRegistrationDescription())` returns an instance of `GcmTemplateRegistrationDescription`

Listing registrations:
- `GetAllRegistrationsAsync`, `GetRegistrationsByChannelAsync`, `GetRegistrationsByTagAsync` returns `Fcm` registration descriptions